File: phpoc-web/test/summary_policy_test.mjs
/**
 * summary_policy_test.mjs — SummaryPolicy test suite.
 *
 * Tests year/month boundary summary block insertion policies:
 *   - YearMonthSummaryPolicy (default): inserts year_summary on year
 *     boundary, month_summary on month boundary
 *   - YearOnlySummaryPolicy: year summary only
 *   - NoSummaryPolicy: never inserts summaries
 *
 * Usage:
 *   node test/summary_policy_test.mjs
 */

import { MockCrypto } from './mock_crypto.mjs';
import { TestHelpers } from './test_helpers.mjs';

const t = new TestHelpers();

// ── Import modules under test ──
let YearMonthSummaryPolicy;
let YearOnlySummaryPolicy;
let NoSummaryPolicy;
try {
  const mod = await import('../src/ledger/summary_policy.js');
  YearMonthSummaryPolicy = mod.YearMonthSummaryPolicy;
  YearOnlySummaryPolicy = mod.YearOnlySummaryPolicy;
  NoSummaryPolicy = mod.NoSummaryPolicy;
} catch (err) {
  YearMonthSummaryPolicy = undefined;
  YearOnlySummaryPolicy = undefined;
  NoSummaryPolicy = undefined;
}

// ── Fixtures ─────────────────────────────────────────────────────────
const MASTER_KEY = 'deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef';
const IDENTITY_SECRET = 'cafebabecafebabecafebabecafebabecafebabecafebabecafebabecafebabe';
const crypto = new MockCrypto();

/**
 * Build a prev_block with the given type, date, and hash values.
 * Supports day, month_summary, and year_summary types.
 */
function makePrevBlock({ type = 'day', date = '2026-01-15', day_hash = null, month_hash = null, year_hash = null, month = null }) {
  const block = { type, date };
  if (day_hash) block.day_hash = day_hash;
  if (month_hash) block.month_hash = month_hash;
  if (year_hash) block.year_hash = year_hash;
  if (month) block.month = month;

  // If no hash provided, compute a deterministic one
  if (!day_hash && !month_hash && !year_hash) {
    block.day_hash = crypto.seal(JSON.stringify(block), MASTER_KEY);
  }

  return block;
}

// ─────────────────────────────────────────────────────────────────────
// Test suite
// ─────────────────────────────────────────────────────────────────────
console.log('\n=== SummaryPolicy Classes Exist ===');

t.assert(typeof YearMonthSummaryPolicy === 'function', 'YearMonthSummaryPolicy is a constructor');
t.assert(typeof YearOnlySummaryPolicy === 'function', 'YearOnlySummaryPolicy is a constructor');
t.assert(typeof NoSummaryPolicy === 'function', 'NoSummaryPolicy is a constructor');

// ── YearMonthSummaryPolicy ────────────────────────────
console.log('\n=== YearMonthSummaryPolicy ===');

if (typeof YearMonthSummaryPolicy === 'function') {
  const policy = new YearMonthSummaryPolicy(crypto, MASTER_KEY);
  const policyWithSig = new YearMonthSummaryPolicy(crypto, MASTER_KEY, IDENTITY_SECRET);

  // Test 1: Same month → no summary blocks
  const prevDay = makePrevBlock({ type: 'day', date: '2026-01-15' });
  const sameMonth = policy.getSummaryBlocks(prevDay, '2026-01-20');
  t.assert(Array.isArray(sameMonth), 'getSummaryBlocks returns an array');
  t.assertEq(sameMonth.length, 0, 'same month returns no summary blocks');

  // Test 2: Month boundary → month_summary block
  const prevJan = makePrevBlock({ type: 'day', date: '2026-01-31' });
  const febEntries = policy.getSummaryBlocks(prevJan, '2026-02-01');
  t.assertEq(febEntries.length, 1, 'month boundary returns 1 summary block');
  t.assertEq(febEntries[0].type, 'month_summary', 'summary block has type "month_summary"');
  t.assertEq(febEntries[0].month, '2026-01', 'month summary covers January');
  t.assert(typeof febEntries[0].month_hash === 'string', 'month summary has month_hash');
  t.assert(/^[0-9a-f]{64}$/.test(febEntries[0].month_hash), 'month_hash is 64 hex chars');

  // Test 3: Month summary block structure
  t.assertHasKeys(febEntries[0], ['type', 'month', 'prev_hash', 'date', 'month_hash'],
    'month_summary has all required keys');
  t.assertEq(febEntries[0].date, '2026-02-01', 'month summary date matches upcoming day');
  t.assertEq(febEntries[0].prev_hash, prevJan.day_hash, 'month summary prev_hash matches prev block');

  // Test 4: Year boundary → year_summary block
  const prevDec = makePrevBlock({ type: 'day', date: '2025-12-31' });
  const jan2026 = policy.getSummaryBlocks(prevDec, '2026-01-01');
  t.assertEq(jan2026.length, 2, 'year boundary returns 2 summary blocks (year + month)');
  t.assertEq(jan2026[0].type, 'year_summary', 'first summary is year_summary');
  t.assertEq(jan2026[0].year, 2025, 'year summary covers 2025');

  // Test 5: Year summary block structure
  t.assertHasKeys(jan2026[0], ['type', 'year', 'prev_hash', 'date', 'year_hash'],
    'year_summary has all required keys');
  t.assertEq(jan2026[0].date, '2026-01-01', 'year summary date matches upcoming day');

  // Test 6: Second block is month summary for December (after year summary)
  t.assertEq(jan2026[1].type, 'month_summary', 'second summary is month_summary');
  t.assertEq(jan2026[1].month, '2025-12', 'month summary covers December');
  t.assertEq(jan2026[1].prev_hash, jan2026[0].year_hash,
    'month summary prev_hash links to year summary');

  // Test 7: Cross-year Dec→Feb (skip a month)
  const prevDecDirect = makePrevBlock({ type: 'day', date: '2025-12-25' });
  const feb2026 = policy.getSummaryBlocks(prevDecDirect, '2026-02-03');
  t.assertEq(feb2026.length, 2, 'Dec→Feb returns 2 summary blocks (year + month)');
  t.assertEq(feb2026[0].type, 'year_summary', 'first is year_summary');
  t.assertEq(feb2026[1].type, 'month_summary', 'second is month_summary');
  t.assertEq(feb2026[1].month, '2026-01', 'month summary covers the skipped January');

  // Test 8: Same year, multiple months gap
  const prevMar = makePrevBlock({ type: 'day', date: '2026-03-15' });
  const mayEntries = policy.getSummaryBlocks(prevMar, '2026-05-01');
  t.assertEq(mayEntries.length, 1, 'Mar→May returns 1 summary block');
  t.assertEq(mayEntries[0].type, 'month_summary', 'summary is month_summary');
  t.assertEq(mayEntries[0].month, '2026-04', 'month summary covers April');

  // Test 9: No summary when prev_block is already a month_summary for same month
  const prevMonthSum = makePrevBlock({
    type: 'month_summary',
    date: '2026-02-01',
    month_hash: crypto.seal('month:2026-01', MASTER_KEY),
    month: '2026-01'
  });
  const afterMonthSum = policy.getSummaryBlocks(prevMonthSum, '2026-02-01');
  t.assertEq(afterMonthSum.length, 0,
    'no summary when prev block is already a month_summary for the same month');

  // Test 10: No redundant December month summary when year summary already covers it
  const prevYearSum = makePrevBlock({
    type: 'year_summary',
    date: '2026-01-01',
    year_hash: crypto.seal('year:2025', MASTER_KEY),
  });
  const afterYearSum = policyWithSig.getSummaryBlocks(prevYearSum, '2026-01-01');
  t.assertEq(afterYearSum.length, 0,
    'no redundant December month summary when year summary is prev block');

  // Test 11: Identity signature on summary blocks
  const prevOct = makePrevBlock({ type: 'day', date: '2025-10-15' });
  const withSigBlocks = policyWithSig.getSummaryBlocks(prevOct, '2026-01-01');
  if (withSigBlocks.length > 0) {
    for (const block of withSigBlocks) {
      t.assert(typeof block.signature === 'string',
        `summary block of type ${block.type} has signature when identity secret is provided`);
    }
  }

  // Test 12: No identity signature without identity secret
  const noSigBlocks = policy.getSummaryBlocks(prevOct, '2026-01-01');
  for (const block of noSigBlocks) {
    t.assertEq(block.signature, undefined,
      `summary block of type ${block.type} has no signature without identity secret`);
  }

  // Test 13: Consecutive month boundaries (multiple month skips)
  const prevApr = makePrevBlock({ type: 'day', date: '2026-04-10' });
  const juneEntries = policy.getSummaryBlocks(prevApr, '2026-06-01');
  t.assertEq(juneEntries.length, 1, 'Apr→Jun returns 1 month summary');
  t.assertEq(juneEntries[0].month, '2026-05', 'month summary covers May (skipped month)');

  // Test 14: Year boundary without month gap (Dec 31 → Jan 1)
  const prevDec31 = makePrevBlock({ type: 'day', date: '2025-12-31' });
  const jan1 = policy.getSummaryBlocks(prevDec31, '2026-01-01');
  t.assertEq(jan1.length, 2, 'Dec 31→Jan 1 returns 2 summaries');
  t.assertEq(jan1[0].type, 'year_summary', 'year summary for 2025');
  t.assertEq(jan1[1].type, 'month_summary', 'month summary for December');
}

// ── YearOnlySummaryPolicy ────────────────────────────
console.log('\n=== YearOnlySummaryPolicy ===');

if (typeof YearOnlySummaryPolicy === 'function') {
  const policy = new YearOnlySummaryPolicy(crypto, MASTER_KEY);

  // Test 15: No month summaries
  const prevJan = makePrevBlock({ type: 'day', date: '2026-01-31' });
  const febOnly = policy.getSummaryBlocks(prevJan, '2026-02-01');
  t.assertEq(febOnly.length, 0, 'YearOnly: month boundary returns no summaries');

  // Test 16: Year boundary returns year summary
  const prevDec = makePrevBlock({ type: 'day', date: '2025-12-31' });
  const janOnly = policy.getSummaryBlocks(prevDec, '2026-01-01');
  t.assertEq(janOnly.length, 1, 'YearOnly: year boundary returns 1 summary');
  t.assertEq(janOnly[0].type, 'year_summary', 'YearOnly: summary is year_summary');
  t.assertEq(janOnly[0].year, 2025, 'YearOnly: year summary covers 2025');

  // Test 17: Same year returns nothing
  const prevMar = makePrevBlock({ type: 'day', date: '2026-03-15' });
  const mayOnly = policy.getSummaryBlocks(prevMar, '2026-05-01');
  t.assertEq(mayOnly.length, 0, 'YearOnly: same year returns no summaries');

  // Test 18: Multiple year gap
  const prev2024 = makePrevBlock({ type: 'day', date: '2024-06-15' });
  const dec2025 = policy.getSummaryBlocks(prev2024, '2026-01-01');
  t.assertEq(dec2025.length, 1, 'YearOnly: multi-year gap returns 1 year summary');
  t.assertEq(dec2025[0].year, 2024, 'YearOnly: year summary covers 2024');
}

// ── NoSummaryPolicy ──────────────────────────────────
console.log('\n=== NoSummaryPolicy ===');

if (typeof NoSummaryPolicy === 'function') {
  const policy = new NoSummaryPolicy(crypto, MASTER_KEY);

  // Test 19: Never returns summaries
  const prev = makePrevBlock({ type: 'day', date: '2025-12-31' });
  const none = policy.getSummaryBlocks(prev, '2026-01-01');
  t.assertEq(none.length, 0, 'NoSummary: year boundary returns nothing');

  // Test 20: Still returns nothing for any boundary
  const prev2 = makePrevBlock({ type: 'day', date: '2026-01-31' });
  const none2 = policy.getSummaryBlocks(prev2, '2026-02-01');
  t.assertEq(none2.length, 0, 'NoSummary: month boundary returns nothing');

  // Test 21: Empty call returns empty array (not null)
  const empty = policy.getSummaryBlocks(prev, '2026-03-01');
  t.assert(Array.isArray(empty), 'NoSummary: always returns an array');
}

// ── Summary ─────────────────────────────────────────────────────────
t.summary('SummaryPolicy');
process.exit(t.failed > 0 ? 1 : 0);

File: phpoc-flutter/test/data/ledger/summary_policy_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:phpoc_flutter/core/crypto/crypto_service.dart';
import 'package:phpoc_flutter/data/ledger/summary_policy.dart';

/// SummaryPolicy — Phase 2 (RED) test suite.
///
/// All 17 assertions from docs/planning/flutter/LEDGER_PHASE1.md Groups L–M:
///   Group L: YearMonthSummaryPolicy (14)
///   Group M: Alternative Policies (3)
///
/// Expected: all tests FAIL (RED) because summary_policy.dart does not exist yet.

// ── Test constants ──────────────────────────────────────────────

const mkHex = '000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f';
const identitySecret = 'identity-secret-32-bytes-xxxxxx';

/// Helper to create a CryptoService and YearMonthSummaryPolicy.
SummaryPolicy _makePolicy({String? identitySecretHex}) {
  final crypto = CryptoService();
  crypto.initialize();
  crypto.setMasterKey(mkHex);
  return YearMonthSummaryPolicy(
    crypto: crypto,
    identitySecret: identitySecretHex,
  );
}

void main() {
  // ═══════════════════════════════════════════════════════════════
  // Group L: YearMonthSummaryPolicy (14 tests)
  // ═══════════════════════════════════════════════════════════════

  group('L: YearMonthSummaryPolicy', () {
    // L1 — getSummaryBlocks returns empty when same month+year as prev
    test(
        'L1: getSummaryBlocks returns empty when same month+year as previous block',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-03-10',
        'day_hash': 'a' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2025-03-15');
      expect(summaries, isEmpty);
    });

    // L2 — inserts month_summary when month changes (same year)
    test(
        'L2: getSummaryBlocks inserts month_summary when month changes (same year)',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-03-31',
        'day_hash': 'b' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2025-04-01');
      expect(summaries, isNotEmpty);
      expect(summaries.first['type'], 'month_summary');
      expect(summaries.first['month'], '2025-03');
    });

    // L3 — inserts year_summary on year boundary, month_summary suppressed (covered by year)
    test(
        'L3: getSummaryBlocks inserts year_summary on year boundary (Dec month covered by year)',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-12-31',
        'day_hash': 'c' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-01-01');
      expect(summaries.length, 1);
      expect(summaries[0]['type'], 'year_summary');
      expect(summaries[0]['year'], 2025);
    });

    // L4 — handles cross-year month gap (Dec→Feb skips Jan)
    test(
        'L4: getSummaryBlocks handles cross-year month gap (Dec→Feb skips Jan)',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-12-15',
        'day_hash': 'd' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-02-01');
      // Should have year_summary(2025) + month_summary(2026-01)
      expect(summaries.length, 2);
      expect(summaries[0]['type'], 'year_summary');
      expect(summaries[1]['type'], 'month_summary');
      expect(summaries[1]['month'], '2026-01');
    });

    // L5 — month_summary has correct month field (YYYY-MM)
    test(
        'L5: getSummaryBlocks month_summary has correct month field (YYYY-MM)',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-07-31',
        'day_hash': 'e' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2025-08-01');
      expect(summaries.first['type'], 'month_summary');
      expect(summaries.first['month'], '2025-07');
    });

    // L6 — year_summary has correct year field (int)
    test('L6: getSummaryBlocks year_summary has correct year field (int)',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-12-31',
        'day_hash': 'f' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-01-01');
      expect(summaries[0]['type'], 'year_summary');
      expect(summaries[0]['year'], 2025);
      expect(summaries[0]['year'], isA<int>());
    });

    // L7 — prev_hash of first summary links to previous block
    test(
        'L7: getSummaryBlocks prev_hash of first summary links to previous block',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-03-31',
        'day_hash': 'abc123' + '0' * 58,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2025-04-01');
      expect(summaries.first['prev_hash'], 'abc123' + '0' * 58);
    });

    // L8 — on year boundary (Dec→Jan), only year_summary; prev_hash links
    test(
        'L8: getSummaryBlocks year_summary prev_hash links to previous block',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-12-31',
        'day_hash': 'g' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-01-01');
      expect(summaries.length, 1);

      // Year summary's prev_hash should match previous block's hash
      expect(summaries[0]['prev_hash'], 'g' * 64);
    });

    // L9 — does not insert year_summary if prev is already year_summary
    test(
        'L9: getSummaryBlocks does not insert year_summary if prev is already year_summary',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'year_summary',
        'year': 2025,
        'date': '2026-01-10',
        'year_hash': 'h' * 64,
        'prev_hash': 'x' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-02-01');
      // Should have month_summary for Jan, but no duplicate year_summary
      final hasYearSummary =
          summaries.any((s) => s['type'] == 'year_summary');
      expect(hasYearSummary, isFalse);

      expect(summaries.length, 1);
      expect(summaries.first['type'], 'month_summary');
    });

    // L10 — does not insert month_summary if prev is already same month_summary
    test(
        'L10: getSummaryBlocks does not insert month_summary if prev is already month_summary for same month',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'month_summary',
        'month': '2025-06',
        'date': '2025-07-01',
        'month_hash': 'i' * 64,
        'prev_hash': 'y' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2025-07-01');
      expect(summaries, isEmpty);
    });

    // L11 — does not insert month summary for December when year summary just inserted
    test(
        'L11: getSummaryBlocks does not insert month summary for December when year summary just inserted',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-12-20',
        'day_hash': 'j' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-01-15');
      // Should have year_summary(2025), but NOT a month_summary for December (2025-12)
      // because the year summary already covers it
      expect(summaries.length, 1);
      expect(summaries[0]['type'], 'year_summary');
    });

    // L12 — month_summary seal (month_hash) is valid
    test('L12: getSummaryBlocks month_summary seal (month_hash) is valid',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-03-31',
        'day_hash': 'k' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2025-04-01');
      final monthBlock = summaries.first;

      expect(monthBlock.containsKey('month_hash'), isTrue);
      expect(monthBlock['month_hash'], isNotEmpty);
      expect(monthBlock['month_hash'].length, 64);
    });

    // L13 — year_summary seal (year_hash) is valid
    test('L13: getSummaryBlocks year_summary seal (year_hash) is valid',
        () {
      final policy = _makePolicy();
      final prevBlock = {
        'type': 'day',
        'date': '2025-12-31',
        'day_hash': 'l' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-01-01');
      final yearBlock = summaries[0];

      expect(yearBlock.containsKey('year_hash'), isTrue);
      expect(yearBlock['year_hash'], isNotEmpty);
      expect(yearBlock['year_hash'].length, 64);
    });

    // L14 — adds identity_seal when identitySecret is set
    test(
        'L14: getSummaryBlocks adds identity_seal when identitySecret is set',
        () {
      final policy = _makePolicy(identitySecretHex: identitySecret);
      final prevBlock = {
        'type': 'day',
        'date': '2025-03-31',
        'day_hash': 'm' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2025-04-01');
      final monthBlock = summaries.first;
      expect(monthBlock.containsKey('identity_seal'), isTrue);
    });
  });

  // ═══════════════════════════════════════════════════════════════
  // Group M: Alternative Policies (3 tests)
  // ═══════════════════════════════════════════════════════════════

  group('M: Alternative Summary Policies', () {
    // M1 — YearOnlySummaryPolicy inserts year_summary but never month_summary
    test(
        'M1: YearOnlySummaryPolicy inserts year_summary but never month_summary',
        () {
      final crypto = CryptoService();
      crypto.initialize();
      crypto.setMasterKey(mkHex);
      final policy = YearOnlySummaryPolicy(crypto: crypto);

      // Month boundary within same year → no summaries (year-only mode)
      final prevBlock = {
        'type': 'day',
        'date': '2025-03-31',
        'day_hash': 'n' * 64,
      };
      var summaries =
          policy.getSummaryBlocks(prevBlock, '2025-04-01');
      expect(summaries, isEmpty);

      // Year boundary → year_summary but no month_summary
      final prevBlock2 = {
        'type': 'day',
        'date': '2025-12-31',
        'day_hash': 'o' * 64,
      };
      summaries = policy.getSummaryBlocks(prevBlock2, '2026-01-01');
      expect(summaries.length, 1);
      expect(summaries[0]['type'], 'year_summary');
    });

    // M2 — YearOnlySummaryPolicy does not insert when prev is year_summary
    test(
        'M2: YearOnlySummaryPolicy does not insert when prev is year_summary',
        () {
      final crypto = CryptoService();
      crypto.initialize();
      crypto.setMasterKey(mkHex);
      final policy = YearOnlySummaryPolicy(crypto: crypto);

      final prevBlock = {
        'type': 'year_summary',
        'year': 2025,
        'date': '2026-01-05',
        'year_hash': 'p' * 64,
        'prev_hash': 'z' * 64,
      };

      final summaries =
          policy.getSummaryBlocks(prevBlock, '2026-02-01');
      expect(summaries, isEmpty);
    });

    // M3 — NoSummaryPolicy never inserts any summary blocks
    test('M3: NoSummaryPolicy never inserts any summary blocks', () {
      final crypto = CryptoService();
      crypto.initialize();
      crypto.setMasterKey(mkHex);
      final policy = NoSummaryPolicy(crypto: crypto);

      // Month boundary
      expect(
        policy.getSummaryBlocks(
            {'type': 'day', 'date': '2025-03-31', 'day_hash': 'q' * 64},
            '2025-04-01'),
        isEmpty,
      );

      // Year boundary
      expect(
        policy.getSummaryBlocks(
            {'type': 'day', 'date': '2025-12-31', 'day_hash': 'r' * 64},
            '2026-01-01'),
        isEmpty,
      );
    });
  });
}
